params/params.go

```diff
@@ -80,9 +82,56 @@ func (v Values) Del(key string) {
 // ensure that an action will not be executed if the query string has been cut
 // off for some reason.
 func (v Values) Encode() string {
+	return v.encode()
+}
+
+// EncodeMultipart returns a ``multipart encoded'' byte buffer
+// with the parameters, along with a Content-Type header string to use,
+// and an error if something somehow goes dramatically wrong.
+func (v Values) EncodeMultipart() (*bytes.Buffer, string, error) {
```

Comment:
It's not so much the inefficiency itself as it is the code cleanliness that bothers me with doing that. I mean, part of it _is_ the inefficiency, don't get me wrong - I don't think it makes a huge deal of sense to convert to a string, then immediately convert back to a `Reader` of some sort - but apart from that, switching data formats constantly in the code just doesn't seem like pretty code to me.

I take your point about the `*bytes.Buffer` being an unnecessary implementation detail, though. If you're happy with doing so, making it an `*io.Reader` seems a reasonable half-way point. (Incidentally, I didn't make that same change because I was worried about potential other applications using Encode() and expecting a string - does it make more sense to change them both and bump the major version?)

Reply:
I think I'd still prefer both encoding methods to return a `string`. If you get a `string` but need an `io.Reader`, all you have to do is call `strings.NewReader(s)`, which doesn't even return an error, so you can do it inline as an expression. If you have an `io.Reader` but need a `string`, that's more complicated (just look at the additional complexity you had to add to the test to do that conversion). Now you have to deal with allocating buffers and checking errors, even though it's all being done in-memory. Of course, it's easy to get a `string` from a `bytes.Buffer`, but then you have to expose the concrete object, which has a huge surface area.

I think we should go with the simplest thing that could possibly work: a `string`. This maintains symmetry with the `Encode` method and avoids breaking compatibility.

Reply:
Also, it looks like the code doesn't compile in its current state. I just tried to enable CI build and test for pull requests. Maybe it'll run it if you push again.

Reply:
Updated accordingly in b82dfc3. Let's see if it works!